Unicode/ustring.c

```diff
@@ -858,6 +862,95 @@ void utf8_strncpy(register char *to, const char *from, int len) {
     to[old-from] = 0;
 }
 
+/* File system encoding functions */
+/**
+ * Converts from file system encoding to unichar_t encoding.
+ * 
+ * @param [out] uto The output buffer
+ * @param [in] from The input buffer.
+ * @param [in] n The number of characters to copy.
+ * @return The output buffer.
+ */
+unichar_t *fsys2u_strncpy(unichar_t *uto, const char *from, size_t n) {
+#ifdef __MINGW32__
+    return utf82u_strncpy(uto, from, n);
+#else
+    return def2u_strncpy(uto, from, n);
+#endif
+}
+
+/**
+ * Converts from unichar_t encoding to file system encoding.
+ * 
+ * @param [out] to The output buffer
+ * @param [in] ufrom The input buffer.
+ * @param [in] n The number of characters to copy.
+ * @return The output buffer.
+ */
+char *u2fsys_strncpy(char *to, const unichar_t *ufrom, size_t n) {
+#ifdef __MINGW32__
+    return u2utf8_strncpy(to, ufrom, n);
+#else
+    return u2def_strncpy(to, ufrom, n);
+#endif
+}
+
+/**
+ * Creates a new buffer, converting from file system encoding to unichar_t.
+ * 
+ * @param [in] from The input buffer
+ * @return The output buffer (NULL on error). Must be freed by the caller.
+ */
+unichar_t *fsys2u_copy(const char *from) {
+#ifdef __MINGW32__
+    return utf82u_copy(from);
+#else
+    return def2u_copy(from);
+#endif
+}
+
+/**
+ * Creates a new buffer, converting from unichar_t to the file system encoding.
+ * 
+ * @param [in] ufrom The input buffer
+ * @return The output buffer (NULL on error). Must be freed by the caller.
+ */
+char *u2fsys_copy(const unichar_t *ufrom) {
+#ifdef __MINGW32__
+    return u2utf8_copy(ufrom);
+#else
+    return u2def_copy(ufrom);
+#endif
+}
+
+/**
+ * Creates a new file system buffer, converting from file system encoding to UTF-8.
+ * 
+ * @param [in] from The input buffer
+ * @return The output buffer (NULL on error). Must be freed by the caller.
+ */
+char *fsys2utf8_copy(const char *from) {
+#ifdef __MINGW32__
+    return copy(from);
+#else
+    return def2utf8_copy(from);
+#endif
+}
+
+/**
+ * Creates a new buffer, converting from UTF-8 to the file system encoding.
+ * 
+ * @param [in] ufrom The input buffer.
+ * @return The output buffer (NULL on error). Must be freed by the caller.
+ */
+char *utf82fsys_copy(const char *ufrom) {
+#ifdef __MINGW32__
+    return copy(ufrom);
+#else
+    return utf82def_copy(ufrom);
+#endif
+}
+
```

Comment:
Yes; tested by essentially anything involving file operations (e.g. opening/closing files) and seeing if they still work. Moreover testing to see if it can open files containing e.g. Cyrillic characters (which it now can).

On anything other than Windows (MinGW, not CygWin), it boils down to calling exactly the same function, so the only potential effect is on Windows. 

Although one thing I've been contemplating is if it makes sense to just use the `2def/def2` functions, and force UTF-8 encoding for Windows.